query/src/main/java/org/infinispan/query/dsl/embedded/impl/SearchQueryMaker.java

```diff
@@ -115,17 +118,48 @@ public SearchQueryParsingResult transform(IckleParsingResult<TypeMetadata> parsi
       indexedEntity = targetedTypeName == null ? searchMapping.indexedEntity(targetedType) :
             searchMapping.indexedEntity(targetedTypeName);
 
-      SearchPredicate predicate = makePredicate(parsingResult.getWhereClause()).toPredicate();
+      InfinispanAggregation<?> aggregation = makeAggregation(scope, parsingResult);
+      SearchPredicate predicate = makePredicate(parsingResult.getWhereClause(), aggregation).toPredicate();
       SearchProjectionInfo projection = makeProjection(parsingResult.getTargetEntityMetadata(), scope.projection(), parsingResult.getProjections(),
-            parsingResult.getProjectedTypes());
+            parsingResult.getProjectedTypes(), aggregation);
       SearchSort sort = makeSort(scope.sort(), parsingResult.getSortFields());
 
-      return new SearchQueryParsingResult(targetedType, targetedTypeName, projection, predicate, sort, hitCountAccuracy);
+      return new SearchQueryParsingResult(targetedType, targetedTypeName, projection, aggregation, predicate, sort, hitCountAccuracy);
+   }
+
+   private <T> InfinispanAggregation makeAggregation(SearchScope<?> scope, IckleParsingResult<TypeMetadata> parsingResult) {
+      PropertyPath[] groupBy = parsingResult.getGroupBy();
+      if (groupBy == null || groupBy.length != 1) {
+         return null;
+      }
+
+      AggregationPropertyPath aggregationPropertyPath = null;
+      Class<T> projectedType = null;
+      boolean displayGroupFirst = false;
+
+      for (int i=0; i<parsingResult.getProjectedPaths().length; i++) {
+         PropertyPath projectedPath = parsingResult.getProjectedPaths()[i];
+         if (projectedPath instanceof AggregationPropertyPath) {
+            if (projectedType != null) {
+               displayGroupFirst = true;
+            }
+            aggregationPropertyPath = (AggregationPropertyPath) projectedPath;
+         } else if (Arrays.equals(groupBy[0].asArrayPath(), projectedPath.asArrayPath())) {
+            projectedType = (Class<T>) parsingResult.getProjectedTypes()[i];
+         }
+      }
+      if (aggregationPropertyPath == null || projectedType == null) {
+         return null;
+      }
+
+      SearchAggregation<? extends Map<T, Long>> searchAggregation = scope.aggregation().terms()
+            .field(groupBy[0].asStringPathWithoutAlias(), projectedType).toAggregation();
```

Comment:
You seem to be ignoring the sort here... You should probably use [`.orderByTermAscending()`](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-order) depending on the query's `order by` clause?

Also, be aware that [the aggregation will return at most `100` entries by default](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-maxtermcount). You might want to tune this or at least document it in Infinispan.

Reply:
> Also, be aware that [the aggregation will return at most 100 entries by default](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-maxtermcount). You might want to tune this or at least document it in Infinispan.

I think I've already tuned it.

> You seem to be ignoring the sort here... You should probably use [.orderByTermAscending()](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-order) depending on the query's order by clause?

I used the sort, it seems to be working

Reply:
Then that's one more things I don't understand, because the very code I commented on doesn't set either option:

```
      SearchAggregation<? extends Map<T, Long>> searchAggregation = scope.aggregation().terms()
            .field(groupBy[0].asStringPathWithoutAlias(), projectedType).toAggregation();
```

Reply:
> I started this contribution from the query from a user that does something like:
> 
> select status, count(code) from org.infinispan.query.model.Sale where day = :day group by status order by status
> @aleoncini could you please confirm it?

yes confirm, the query well represents the use case of the customer. HTH